src/net/socket.cpp

```diff
@@ -0,0 +1,282 @@
+#include <net/socket.hpp>
+
+namespace arch::net {
+	const IPv4 Socket::any_address(INADDR_ANY);
+	const Socket::port_type Socket::random_port = 0;
+
+	Socket::Socket(protocol_t p) {
+		switch (p) {
+			case protocol_t::UDP:
+				_socket = socket(AF_INET, SOCK_DGRAM, IPPROTO_UDP);
+				break;
+			case protocol_t::TCP:
+				_socket = socket(AF_INET, SOCK_STREAM, IPPROTO_TCP);
+				break;
+			default:
+				// log error
+				break;
+		}
+
+		if (_socket == INVALID_SOCKET) {
+			// log error
+		}
+
+		_proto = p;
+	}
+	Socket::Socket(protocol_t p, IPv4 address, port_type port) :
+		Socket(p) {
+		if (not bind(address, port)) {
+			// log error
+		}
+	}
+	Socket::Socket(protocol_t p, port_type port) :
+		Socket(p) {
+		if (not bind(port)) {
+			// log error
+		}
+	}
+	Socket::~Socket() {
+		close();
+	}
+
+	void Socket::close() {
+		if (_socket != INVALID_SOCKET) {
+			if (
+#ifdef _WIN32
```

Comment:
Having windows and linux implementations in one place using ifdefs all over the place is pretty spaghetti if you ask me. In my opinion, it should be refactored and separated. For example, there could a base class with common code, and two classes with windows and unix implemenations respectively. This way we:
- wouldn't have to rely on preprocessor macros everywhere
- if something is wrong with one implementation, we immediately know where to jump to
- person working with one implementation does not have to look at code from other implementation

Reply:
Something like this?:

impl.cpp:
```c++
void _class::foo(){
    ...
}
#ifdef WIN32
    // #include windows_impl.cpp or something
#elif defined unix
    // #include unix_impl.cpp or something
#endif
void _class::bar(){
    ...
}
```

Reply:
Is it really worth it?
Most times its here just to switch between single lines.
Depending on OS the other part is basically a comment. FInding an issue wouldn't be a problem.